include/nbl/system/ISystem.h

```diff
@@ -0,0 +1,244 @@
+#ifndef __NBL_I_SYSTEM_H_INCLUDED__
+#define __NBL_I_SYSTEM_H_INCLUDED__
+
+#include <variant>
+#include "nbl/core/IReferenceCounted.h"
+#include "nbl/system/ISkippableAsyncQueueDispatcher.h"
+#include "nbl/system/IFileArchive.h"
+#include "nbl/system/IWriteFile.h"
+
+namespace nbl {
+namespace system
+{
+
+class ISystem final : public core::IReferenceCounted
+{
+public:
+    class ISystemCaller : core::IReferenceCounted
+    {
+    protected:
+        virtual ~ISystemCaller() = default;
+
+    public:
+        virtual core::smart_refctd_ptr<IReadFile> createReadFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual core::smart_refctd_ptr<IWriteFile> createWriteFile(ISystem* sys, const std::filesystem::path& filename) = 0;
+        virtual uint32_t read(IReadFile* file, void* buffer, size_t offset, size_t size) = 0;
+        virtual uint32_t write(IWriteFile* file, const void* buffer, size_t offset, size_t size) = 0;
+    };
+
+protected:
+    struct SRequestType;
+    class CAsyncQueue;
+
+    class SFutureBase
+    {
+        friend ISystem;
+
+    protected:
+        bool valid_flag = false;
+        SRequestType* request;
+        CAsyncQueue* asyncQ;
+
+    public:
+        // SFuture is non-copyable and non-movable
+        SFutureBase(const SFutureBase&) = delete;
+
+        ~SFutureBase()
+        {
+            request->set_skip();
+        }
+
+        bool ready() const { return !request || request->ready; }
+        bool valid() const { return valid_flag; }
+
+        void wait()
+        {
+            asyncQ->waitForRequestCompletion(request[0]);
+        }
+    };
+
+    template <typename T>
+    class SFuture : public SFutureBase
+    {
+        alignas(T) uint8_t storage[sizeof(T)];
```

Comment:
Lots of memory will be wasted because of the aligned type not being the first member.

inheritance makes this complicated, but you should have a `SFutureStorage` with this array and inherit from that before the `SFutureBase`

Reply:
Actually this is VERY important, if you dont inherit from `SFutureStorage` then the destructor of `SFuture` gets called, meaning that the storage's byte array ends its lifetime (another variable from the stack can start life there) before the destructor of `SFutureBase` is called.

This could be a problem when `valid_flag=false` because even though you havent constructed an object there and wont call a destructor, the lifetime of the `storage` array is finished while the request has not notified yet. In the case of the `~SFuture` being already executed but the request being processed just before the `~SFutureBase` calls the `abort`/`cancel`, the request may write to the storage which has now ended its lifetime.

So making sure the `storage` ends its lifetime after a cancellation/abortion is crucial (and also making sure that a cancellation/abortion cannot happen midway processing of the request, because you check the predicate before you start processing, and while you're processing you expect to have a return value being alive to be able to write to)